<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [3]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [4]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [5]:
query_3_1 = f'''select 
           count(*)
           FROM public.vacancies
'''


In [6]:
df = pd.read_sql_query(query_3_1, connection)
df.iloc[0][0]

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\1782510478.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [7]:
query_3_2 = f'''select 
           count(*)
           FROM public.employers
'''


In [8]:
df = pd.read_sql_query(query_3_2, connection)
df.iloc[0][0]

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\414079917.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [9]:
query_3_3 = f'''select 
           count(*)
           FROM public.areas
'''


In [10]:
df = pd.read_sql_query(query_3_3, connection)
df.iloc[0][0]

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\162519764.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [11]:
query_3_4 = f'''select 
           count(*)
           FROM public.industries
'''


In [12]:
df = pd.read_sql_query(query_3_4, connection)
df.iloc[0][0]

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\2325475411.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


294

***

- Количесвто областей 1362 
- Количество сфер деятельности 294
- Количество вакнсий ~ 50000 в тоже время только работодателей 23500, примерное соотношениие 1 к 2 но распеределение количества вакансий на каждого работодателя очень размашестое выведим топ 10 работодателей по кол. вакансий .
-  Сразу видно не равномерное распределений вакансий если с 2 по 10 изменения в районе 25% то яндекс больше сразу в 4 раза.

In [13]:
query_3_v = f'''select 
           e.name,
           count(v.id)
           From public.vacancies v 
           left join public.EMPLOYERS e on e.id = v.employer_id
           group by e.name
           order by 2 desc
           limit 10
'''
df = pd.read_sql_query(query_3_v, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\1402077674.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_v, connection)


name  count
0                 Яндекс   1933
1             Ростелеком    491
2               Тинькофф    444
3                   СБЕР    428
4          Газпром нефть    331
5            ИК СИБИНТЕК    327
6                    МТС    292
7                DataArt    247
8  Совкомбанк Технологии    204
9             Первый Бит    176

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [14]:
query_4_1 = f'''select 
           a.name area,
           count(v.id) cnt
           FROM public.areas a
           left join public.vacancies v on v.area_id = a.id
           group by a.id
           order by 2 desc
           limit 5
'''

In [15]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [16]:
query_4_2 = f'''select 
            count(*)
            FROM public.vacancies 
            where salary_from is not null or salary_to is not null
'''

In [17]:
df = pd.read_sql_query(query_4_2, connection)
df.iloc[0][0]

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\3638440477.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [18]:
query_4_3 = f'''select 
            'Нижняя граница:' as name ,
            round(avg(salary_from)) avg
            FROM public.vacancies 
            union all
            select 
            'Верхняя граница:' as  name ,
            round(avg(salary_to)) avg
            FROM public.vacancies 
'''

In [19]:
df = pd.read_sql_query(query_4_3, connection)

df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\1470339711.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


name       avg
0   Нижняя граница:   71065.0
1  Верхняя граница:  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [20]:
query_4_4 = f'''select 
           schedule,
           employment,
           count(employment)
           From public.vacancies 
           group by schedule, employment
           order by 3 desc
'''

In [21]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [22]:
query_4_5 = f'''select 
           experience,
           count(experience)
           From public.vacancies 
           group by experience
           order by 2 
'''

In [23]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

- Как видно 10% и 5% от всех вакансий находятся в Москве и Санкт-Петербург. 
- Не понятен следущий факт что тлько у 24073 вакансий указана ЗП, что является меньше половины от общего кол. можно предположитть что работодатель во время собеседования расчитывает на неуверености кондидата уменьшить ЗП. так как она не указана с самого начала.
- средняя ЗП нижней и верхней граници возможно не имеет смысл так каак в столице и в питере на уровень выше ЗП чем у остольных регионов. 
- Полный день полная занятость №1 естествено так как работодатели чаще ищат на постаянку, радует количество удаленной работы ~ 16% а если с частичной занятостью то ~ 18%
- что с опытом то тут все все не однозначно. Половина всех вакансий от 1 года до 3 возможно связано с поиском самымиых молодымых специалистов но имеющиеми какойто опыт , без опыта вакансий ~4 раза меньше, от 3 до 6 ищут реже в 2 раза чем 1 до 3 , более 6 лет всего 1337 можно сказать чем больше опыт тем меньше вакансий если не считать без опыта

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
query_5_1 = f'''(select 
           e.name,
           count(v.id)
           From public.vacancies v 
           left join public.EMPLOYERS e on e.id = v.employer_id
           group by e.name
           order by 2 desc
           limit 1)
           union all 
           (select 
           e.name,
           count(v.id)
           From public.vacancies v 
           left join public.EMPLOYERS e on e.id = v.employer_id
           group by e.name
           order by 2 desc
           offset 4
           limit 1)
'''

In [25]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
query_5_2 = f'''select 
           a.name,
           count(e.id),
           count(v.id)
           From public.areas a
           left join public.vacancies v  on a.id = v.area_id
           left join public.EMPLOYERS e  on a.id = e.area
           group by a.id
           having count(v.id) = 0
           order by 2 desc 
           limit 1
'''

In [27]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  count  count
0  Россия    410      0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
query_5_3 = f'''select 
           e.name,
           count(distinct a.name)
           From public. vacancies v
           left join public.EMPLOYERS e  on e.id = v.employer_id
           left join public.areas a  on a.id = v.area_id
           group by e.name 
           order by 2 desc
'''

In [29]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
query_5_4 = f'''select 
            count(*)
            From public. EMPLOYERS e
            left join public.EMPLOYERS_INDUSTRIES ei on e.id = ei.employer_id
            where employer_id is null
'''

In [31]:
df = pd.read_sql_query(query_5_4, connection)
df.iloc[0][0]

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\352366383.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
query_5_5 = f'''select 
            e.name,
            count(i.name)
            From public. EMPLOYERS e
            full join public.EMPLOYERS_INDUSTRIES ei on e.id = ei.employer_id
            full join public.INDUSTRIES i on ei.industry_id = i.id
            group by e.id
            having count(i.name) = 4
            order by 1
            offset 2 
            limit 1
'''

In [33]:
df = pd.read_sql_query(query_5_5, connection)
df.iloc[0][0]

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\307362801.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


'2ГИС'

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
query_5_6 = f'''select 
            count(*)
            From public. EMPLOYERS e
            full join public.EMPLOYERS_INDUSTRIES ei on e.id = ei.employer_id
            left join public.INDUSTRIES i on ei.industry_id = i.id
            where i.name = 'Разработка программного обеспечения'
'''

In [35]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
import urllib.parse as req
import requests
from bs4 import BeautifulSoup

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

tables = pd.read_html(url)
df = tables[1]
cities = list(df['Город'])
cities = tuple(cities)
print(cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [37]:
query_5_7 = f'''select 
            a.name,
            count(a.name)
            from vacancies v
            left join employers e on v.employer_id = e.id
            left join areas a on a.id = v.area_id
            where e.name = 'Яндекс'
            and a.name in {cities}
            group by a.name
            
            union all
            
            select 
            'total',
            count(*)
            from vacancies v
            left join employers e on v.employer_id = e.id
            left join areas a on a.id = v.area_id
            where e.name = 'Яндекс'
            and a.name in {cities}
            order by 2
            
'''


In [38]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            total    485

***

- большой охват по поиску сотрудников т.е больше чем в 100 регионах это (Яндекс, Ростелеком, Спецремонт)
- примерно 1/3 работодателей не указали сферу деятельности. 
- примерно 13% работодателей занимаются разработкой ПО

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [39]:
query_6_1 = f'''select 
            count(*)
            From public.vacancies
            where  lower(name) like '%data%' 
            or  lower(name) like '%данн%'
'''

In [40]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [41]:
query_6_2 = f'''select 
            count(*)
            From public.vacancies
            where  (lower(name) like '%junior%'
            or  experience = 'Нет опыта'
            or employment = 'Стажировка')
            and (lower(name) like '%data scientist%' 
            or  lower(name) like '%исследователь данных%'
            or lower(name) not like '%html%'
            and lower(name) like '%ml%'
            or  lower(name) like '%machine learning%'
            or  lower(name) like '%машинн%обучен%')
            
'''

In [42]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [43]:
query_6_3 = f'''select 
            count(*)
            From public.vacancies
            where (lower(name) like '%data%scientist%'
            or lower(name) like '%data%science%'
            or lower(name) like '%исследователь%данных%'
            or lower(name) like '%machine%learning%'
            or lower(name) like '%машинн%обучен%'
            or lower(name) not like '%html%'
            and name like '%ML%')
            and
            (lower(key_skills) like '%sql%'
            or lower(key_skills) like '%postgres%')
'''


In [44]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [45]:
query_6_4 = f'''select 
            key_skills
            From public.vacancies
            where (lower(name) like '%data%scientist%'
            or lower(name) like '%data%science%'
            or lower(name) like '%исследователь%данных%'
            or lower(name) like '%machine%learning%'
            or lower(name) like '%машинн%обучен%'
            or lower(name) not like '%html%'
            and name like '%ML%')
            and
            lower(key_skills) like '%python%'
'''


In [46]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


key_skills
0            Python\tSQL\tРабота в команде\tOpenCV\tML
1    Oracle Pl/SQL\tБазы данных\tMachine Learning\t...
2    Python\tOpenCV\tTensorflow\tComputer Vision\tН...
3                       Python\tFlask\tLinux\tAWS\tGit
4     Python\tSQL\tNLP\tScikit-learn\tMachine Learning
..                                                 ...
346  Python\tLinux\tSQL\tPyTorch\tOpenCV\tKeras\tCa...
347  Python\tАналитическое мышление\tАнализ данных\...
348  Python\tNumpy\tPyTorch\tPandas\tМатематический...
349           Python\tScikit-learn\tPandas\tGit\tNumpy
350  Python\tLinux\tKubernetes\tAWS\tGit\tPeople Ma...

[351 rows x 1 columns]

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [47]:
query_6_5 = f'''select 
            round(avg(length(key_skills)
                -length(replace(key_skills, CHR(9), ''))+1), 2)
            From public.vacancies
            where (lower(name) like '%data%scientist%'
            or lower(name) like '%data%science%'
            or lower(name) like '%исследователь%данных%'
            or lower(name) like '%machine%learning%'
            or lower(name) like '%машинн%обучен%'
            or lower(name) not like '%html%'
            and name like '%ML%')

'''


In [48]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\390248827.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [49]:
query_6_6 = f'''select 
            distinct experience,
            round(avg(coalesce((salary_from+salary_to)/2,salary_from,salary_to)))
            From public.vacancies
            where (lower(name) like '%data%scientist%'
            or lower(name) like '%data%science%'
            or lower(name) like '%исследователь%данных%'
            or lower(name) like '%machine%learning%'
            or lower(name) like '%машинн%обучен%'
            or lower(name) not like '%html%'
            and name like '%ML%')
            group by experience

'''

In [50]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\3561045109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  139675.0
3       От 3 до 6 лет  243115.0

***

- Большие ЗП у сантистов что тут сказать еще ).
- 3.4% из всех вакансий затрагивают данные
- Вакансий с SQL всего 200 а вакансий с данными 1700 на чем же они работают?
- Python нужен для 350 вакансий
- Работодатель требует в средне больше 6 навыков , чем больше ЗП тем больше скилов нужно
- ЗП растут ~ в 2 раза за 3 года , так как професия еще новая людей с стажом больше 6 лет мало

# Общий вывод по проекту

- На первый взгляд количество вакансий 50000 что кажется весьма обшироной БД но после выборок особо ни чего не остается , большая вероятность выбросов , и искажения данных. Тут уже понимаеш почему БД на гигабайты так как на малых данных особо не расчитываеш.
- Распеределение кол. вакансий по областямне равномерное так как и плотность населения соответственно лидируют Москва и Санкт-Петербург 
- Количество Вакансий удаленной работы почти 20% весьма внушительная цифраю
- По работо дателям можно сказать что Яндекс топ 1 компания России в сфере IT так и еще производит набор почти во всех регионах и делает это больше чем другие компании в разы
- По DS сказать особо не чего, всего 450 вакансий требуют как SQL и Python ну и ЗП космические так же как и требования 

### Узнаем с какими навыками требуются DS.
1. Сделаем запрос всех навыков DS вакансий.
2. Удалим пропуски 
3. Создадим функцию которая :
- примет датафрейм 
- переберет занчения через цикл 
- добавит в словарь ключь это название навыка и значение это количество этого навыка в вакансиях
- если навык уже в славаре то значение увеливается на 1 
- после переводим в датафрейм и выводим его 
4. Выведим топ 10 навыков

In [51]:
query_o = f'''select 
            replace(key_skills, CHR(9), '|')
            From public.vacancies
            where lower(name) like '%data%scientist%'
            or lower(name) like '%data%science%'
            or lower(name) like '%исследователь%данных%'
            or lower(name) like '%machine%learning%'
            or lower(name) like '%машинн%обучен%'
            or lower(name) not like '%html%'
            and name like '%ML%'
'''

In [64]:
df1 = pd.read_sql_query(query_o, connection)
df1 = df1.dropna(thresh=1, axis=0) # удаляем пропуски
def dan(args):
    dicter = dict()                       # создаем пустой словарь
    lines = df1.shape[0]                  # вычесляем длину датафрейма
    for i in range(0,lines):              # запускаем цикл
        arg = args.iloc[i][0]             # берем строку с навыками из DF
        result =  arg.split('|')          # преобразуем его в список навыков
        for j in result:                  # перебераем список
            if j in dicter.keys():        # если навык уже в словаре то увеличиваем значение на 1
                dicter[j] = dicter[j] + 1
            else:                         # если не то вносим его в словарь со значением 1
                dicter.update({j:1})
    value = pd.Series(dicter, name = 'count')
    value = pd.DataFrame(value)           # преобразуем в DF
    return value


a = dan(df1)
a = a.sort_values(by='count',ascending=False) # сортируем и выводим топ 10 навыков
a.head(10)

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_15460\3847574075.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df1 = pd.read_sql_query(query_o, connection)


count
Python                       348
SQL                          191
Machine Learning             114
Git                           66
Математическая статистика     63
Data Analysis                 54
Data Science                  52
Pandas                        52
ML                            49
Английский язык               48

- Как видно Python топ 1 естественно так как пандас и нампи в нем, 
- SQL самая распростроненная СУБД . 
- ML большие данные в ручную не кто не хочет обрабатывать,
- Английский радует